x/data/server/lookup/table.go

```diff
@@ -0,0 +1,159 @@
+package lookup
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"hash/fnv"
+)
+
+// Table is the interface for interacting with a lookup table.
+type Table interface {
+	// GetOrCreateID is an idempotent method for creating or retrieving an unique
+	// shortened identifier for the provided binary value.
+	GetOrCreateID(store KVStore, value []byte) []byte
+
+	// GetValue returns the binary data (if any) corresponding to the provided shortened identifier.
+	GetValue(store KVStore, id []byte) []byte
+}
+
+// NewTable creates a new lookup table for the provided, optional KVStore prefix using default parameters.
+// Default parameters are currently set to use the first 4-bytes of the FNV-1a 64-bit, non-cryptographic hash.
+// In the case of a collision, more bytes of the hash will be used for disambiguation but this happens
+// in a minority of cases except for massively large data sets.
+func NewTable(prefix []byte) (Table, error) {
+	return NewTableWithOptions(TableOptions{
+		Prefix: prefix,
+	})
+}
+
+// NewTableWithOptions creates a Table with custom options. Most users should just use NewTable
+// with the default values.
+func NewTableWithOptions(options TableOptions) (Table, error) {
+	prefixLen := len(options.Prefix)
+	minLength := options.MinLength
+	if minLength == 0 {
+		minLength = 4
+	}
+
+	newHash := options.NewHash
+	if newHash == nil {
+		newHash = func() hash.Hash {
+			return fnv.New64a()
+		}
+	}
+
+	hashLen := len(newHash().Sum(nil))
+	if minLength > hashLen {
+		return nil, fmt.Errorf("option MinLength %d is greater than hash length %d", minLength, hashLen)
+	}
+
+	bufLen := prefixLen + hashLen + binary.MaxVarintLen64
```

Comment:
Also, I think `4` bytes are enough for disambiguation bytes - no need for 10 bytes - that's more then `uint64`.

Reply:
In testing, I never came across a scenario where more than 6 total bytes were needed with 10^8 cases so the algorithm is overly conservative and always allocates a buffer which is too big. I do want it to be theoretically impossible to have a collision which is why the varint is appended in the end, but with any decent hash function and proper parameters it will _never_ be needed.

Reply:
Correct. I feel we should update the documentation. Now it says that the minimum ID length is `4` bytes. But it should be `prefixLen + minHashLen`.

Reply:
The minimum ID length is 4 bytes, it just might be stored with a prefix in the KVStore, but that doesn't change the ID length